File: XSNetwork/Classes/XSCustomResponseSerializer.m
//
//  XSCustomResponseSerializer.m
//  XSNetwork
//
//  Created by shun on 2018/1/16.
//  Copyright © 2018年 xiangshun. All rights reserved.
//

#import "XSCustomResponseSerializer.h"

static NSString * const XSURLResponseSerializationErrorDomain = @"com.xsnetwork.error.serialization.response";

static id XSJSONObjectByRemovingKeysWithNullValues(id JSONObject, NSJSONReadingOptions readingOptions) {
    if ([JSONObject isKindOfClass:[NSArray class]]) {
        NSMutableArray *mutableArray = [NSMutableArray arrayWithCapacity:[(NSArray *)JSONObject count]];
        for (id value in (NSArray *)JSONObject) {
            [mutableArray addObject:XSJSONObjectByRemovingKeysWithNullValues(value, readingOptions)];
        }
        return (readingOptions & NSJSONReadingMutableContainers) ? mutableArray : [NSArray arrayWithArray:mutableArray];
    } else if ([JSONObject isKindOfClass:[NSDictionary class]]) {
        NSMutableDictionary *mutableDictionary = [NSMutableDictionary dictionaryWithDictionary:JSONObject];
        for (id <NSCopying> key in [(NSDictionary *)JSONObject allKeys]) {
            id value = (NSDictionary *)JSONObject[key];
            if (!value || [value isEqual:[NSNull null]]) {
                [mutableDictionary removeObjectForKey:key];
            } else if ([value isKindOfClass:[NSArray class]] || [value isKindOfClass:[NSDictionary class]]) {
                mutableDictionary[key] = XSJSONObjectByRemovingKeysWithNullValues(value, readingOptions);
            }
        }
        return (readingOptions & NSJSONReadingMutableContainers) ? mutableDictionary : [NSDictionary dictionaryWithDictionary:mutableDictionary];
    }
    return JSONObject;
}


@implementation XSCustomResponseSerializer

+ (instancetype)serializer {
    return [self serializerWithReadingOptions:(NSJSONReadingOptions)0];
}

+ (instancetype)serializerWithReadingOptions:(NSJSONReadingOptions)readingOptions {
    XSCustomResponseSerializer *serializer = [[self alloc] init];
    serializer.readingOptions = readingOptions;
    return serializer;
}

- (instancetype)init {
    self = [super init];
    if (!self) {
        return nil;
    }

    self.acceptableStatusCodes = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(200, 100)];
    self.acceptableContentTypes = [NSSet setWithObjects:
                                   @"application/json",
                                   @"text/json",
                                   @"text/javascript",
                                   @"text/plain",
                                   @"text/html",
                                   nil];
    return self;
}

#pragma mark - Validation

- (BOOL)validateResponse:(NSHTTPURLResponse *)response
                    data:(NSData *)data
                   error:(NSError **)error {
    if (!response) {
        return YES;
    }

    if (![self.acceptableStatusCodes containsIndex:(NSUInteger)response.statusCode]) {
        NSString *description = [NSHTTPURLResponse localizedStringForStatusCode:response.statusCode];
        NSDictionary *userInfo = @{
            NSLocalizedDescriptionKey: description,
            NSURLErrorFailingURLErrorKey: response.URL ?: [NSNull null]
        };
        if (error) {
            *error = [NSError errorWithDomain:NSURLErrorDomain
                                         code:response.statusCode
                                     userInfo:userInfo];
        }
        return NO;
    }

    if (response.MIMEType && self.acceptableContentTypes &&
        ![self.acceptableContentTypes containsObject:response.MIMEType]) {
        NSString *description = [NSString stringWithFormat:
                                 @"Request failed: unacceptable content-type: %@", response.MIMEType];
        NSDictionary *userInfo = @{
            NSLocalizedDescriptionKey: description,
            NSURLErrorFailingURLErrorKey: response.URL ?: [NSNull null]
        };
        if (error) {
            *error = [NSError errorWithDomain:XSURLResponseSerializationErrorDomain
                                         code:NSURLErrorCannotDecodeContentData
                                     userInfo:userInfo];
        }
        return NO;
    }

    return YES;
}

#pragma mark - Response Object

- (id)responseObjectForResponse:(NSURLResponse *)response
                           data:(NSData *)data
                          error:(NSError **)error {
    NSError *validationError = nil;
    if (![self validateResponse:(NSHTTPURLResponse *)response data:data error:&validationError]) {
        if (error) {
            *error = validationError;
        }
        return nil;
    }

    // Treat a single space (Rails head :ok workaround) as empty response
    BOOL isSpace = [data isEqualToData:[NSData dataWithBytes:" " length:1]];
    if (data.length == 0 || isSpace) {
        return nil;
    }

    NSError *serializationError = nil;
    id responseObject = [NSJSONSerialization JSONObjectWithData:data
                                                        options:self.readingOptions
                                                          error:&serializationError];

    if (self.removesKeysWithNullValues && responseObject) {
        responseObject = XSJSONObjectByRemovingKeysWithNullValues(responseObject, self.readingOptions);
    }

    if (serializationError && error) {
        *error = serializationError;
    }

    return responseObject;
}

#pragma mark - NSCopying

- (instancetype)copyWithZone:(NSZone *)zone {
    XSCustomResponseSerializer *serializer = [[[self class] allocWithZone:zone] init];
    serializer.readingOptions = self.readingOptions;
    serializer.removesKeysWithNullValues = self.removesKeysWithNullValues;
    serializer.acceptableStatusCodes = [self.acceptableStatusCodes copy];
    serializer.acceptableContentTypes = [self.acceptableContentTypes copy];
    return serializer;
}

@end

File: XSNetwork/Classes/XSCustomResponseSerializer.swift
//  XSCustomResponseSerializer.swift
//  XSNetwork
import Foundation

@objcMembers
@objc public class XSCustomResponseSerializer: NSObject, NSCopying {
    public var readingOptions: JSONSerialization.ReadingOptions = []
    public var removesKeysWithNullValues: Bool = false
    public var acceptableStatusCodes: IndexSet = IndexSet(integersIn: 200..<300)
    public var acceptableContentTypes: Set<String> = ["application/json","text/json","text/javascript","text/plain","text/html"]

    public required override init() { super.init() }

    @objc public static func serializer() -> Self { return serializerWithReadingOptions([]) }

    @objc public static func serializerWithReadingOptions(_ options: JSONSerialization.ReadingOptions) -> Self {
        let s = self.init(); s.readingOptions = options; return s
    }

    @objc public func responseObject(for response: URLResponse?, data: Data?, error: AutoreleasingUnsafeMutablePointer<NSError?>?) -> Any? {
        if let http = response as? HTTPURLResponse, !acceptableStatusCodes.contains(http.statusCode) {
            error?.pointee = NSError(domain: NSURLErrorDomain, code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: HTTPURLResponse.localizedString(forStatusCode: http.statusCode)])
            return nil
        }
        guard let d = data, !d.isEmpty, !(d.count == 1 && d[0] == 0x20) else { return nil }
        do {
            var obj = try JSONSerialization.jsonObject(with: d, options: readingOptions)
            if removesKeysWithNullValues { obj = removeNulls(obj) }
            return obj
        } catch let e as NSError { error?.pointee = e; return nil }
    }

    private func removeNulls(_ obj: Any) -> Any {
        if let arr = obj as? [Any] { return arr.map { removeNulls($0) } }
        if let dict = obj as? [AnyHashable: Any] {
            var r = dict
            for k in dict.keys {
                if dict[k] == nil || dict[k] is NSNull { r.removeValue(forKey: k) }
                else if let v = dict[k] { r[k] = removeNulls(v) }
            }
            return r
        }
        return obj
    }

    public func copy(with zone: NSZone? = nil) -> Any {
        let c = XSCustomResponseSerializer(); c.readingOptions = readingOptions
        c.removesKeysWithNullValues = removesKeysWithNullValues
        c.acceptableStatusCodes = acceptableStatusCodes; c.acceptableContentTypes = acceptableContentTypes
        return c
    }
}
